curp/src/client_new/retry.rs

```diff
@@ -0,0 +1,255 @@
+#![allow(unused)] // TODO: remove
+
+use std::{marker::PhantomData, ops::SubAssign, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use curp_external_api::cmd::Command;
+use futures::Future;
+use tracing::warn;
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        connect::ConnectApi, ConfChange, CurpError, FetchClusterResponse, Member, ReadState,
+        Redirect,
+    },
+};
+
+use super::{ClientApi, LeaderStateUpdate, ProposeResponse};
+
+/// Backoff config
+#[derive(Debug, Clone)]
+pub(super) enum BackoffConfig {
+    /// A fixed delay backoff
+    Fixed,
+    /// A exponential delay backoff
+    Exponential {
+        /// Control the max delay of exponential
+        max_delay: Duration,
+    },
+}
+
+/// Retry config to control the retry policy
+#[derive(Debug, Clone)]
+pub(super) struct RetryConfig {
+    /// Backoff config
+    backoff: BackoffConfig,
+    /// Initial delay
+    delay: Duration,
+    /// Retry count
+    count: usize,
+    /// Enable jitter to randomize the delay to avoid thundering herd
+    jitter: bool,
+}
+
+/// Backoff tool
+#[derive(Debug)]
+struct Backoff {
+    /// The retry config
+    config: RetryConfig,
+    /// Current delay
+    cur_delay: Duration,
+}
+
+impl RetryConfig {
+    /// Create a fixed retry config
+    fn new_fixed(delay: Duration, count: usize, jitter: bool) -> Self {
+        assert!(count > 0, "retry count should be larger than 0");
+        Self {
+            backoff: BackoffConfig::Fixed,
+            delay,
+            count,
+            jitter,
+        }
+    }
+
+    /// Create a exponential retry config
+    fn new_exponential(delay: Duration, max_delay: Duration, count: usize, jitter: bool) -> Self {
+        assert!(count > 0, "retry count should be larger than 0");
+        Self {
+            backoff: BackoffConfig::Exponential { max_delay },
+            delay,
+            count,
+            jitter,
+        }
+    }
+
+    /// Create a backoff process
+    fn init_backoff(&self) -> Backoff {
+        Backoff {
+            config: self.clone(),
+            cur_delay: self.delay,
+        }
+    }
+}
+
+impl Backoff {
+    /// Get the next delay duration, None means the end.
+    fn next_delay(&mut self) -> Option<Duration> {
+        if self.config.count == 0 {
+            return None;
+        }
+        self.config.count.sub_assign(1);
+        let mut cur = self.cur_delay;
+        if let BackoffConfig::Exponential { max_delay } = self.config.backoff {
+            self.cur_delay = self
+                .cur_delay
+                .checked_mul(2)
+                .unwrap_or(self.cur_delay)
+                .min(max_delay);
+        }
```

Comment:
`config` should be read-only, and we'd better create a writable `count` like `cur_delay`, rather than just decrease `count` in `config`

Reply:
done